File: domain/src/main/java/com/oe/ziel/domain/booking/solver/Grid.java
package com.oe.ziel.domain.booking.solver;

import org.joda.time.Duration;
import org.joda.time.Instant;
import org.joda.time.Period;

public class Grid {

    private final Integer steps;
    private final Instant start;
    private final Instant end;
    private final Period period;
    private final Duration duration;

    public Grid(Integer steps, Instant start, Instant end) {
        this.steps = steps;
        this.start = start;
        this.end = end;
        this.period = new Period(start, end);
        this.duration = new Duration(start, end);
    }

    /** Returns [0, 1] where the allocation sits on the entire spectrum*/
    public Double t(Integer i) {
        return i.doubleValue() / steps;
    }

    /** Returns the duration from the start time */
    public Duration h(Integer i) {
        return duration.multipliedBy(i).dividedBy(steps);
    }

    /** Returns an instant at Step i */
    public Instant instant(Integer i) {
        return start.plus(h(i));
    }

    public Integer getSteps() {
        return steps;
    }

    public Instant getStart() {
        return start;
    }

    public Instant getEnd() {
        return end;
    }

    public Period getPeriod() {
        return period;
    }

    public Duration getDuration() {
        return duration;
    }
}

File: legacy-planning-domain/src/main/scala/com/oe/ziel/domain/grid/Grid.scala
package com.oe.ziel.domain.grid

import com.oe.ziel.DEFAULTS
import com.oe.ziel.domain.Work
import org.joda.time.{Duration, Instant, Period}



sealed trait Grid {
  def steps : Int
  def start: Instant
  def end: Instant

  // concrete methods
  def period = new Period(start, end)
  def duration = new Duration(start, end)
  def iterable : Iterable[Int] = (0 to steps).toStream // TODO - can do allot more here

  /** Returns [0, 1] where the allocation sits on the entire spectrum*/
  def t(i : Int) = i.toDouble / steps
  /** Returns the duration from the start time */
  def h(i : Int) = duration.multipliedBy(i).dividedBy(steps)
  /** Returns an instant at Step i */
  def instant(i : Int) = start.plus(h(i))
}


case class GridStepEachHour(
                         start : Instant,
                         end : Instant) extends Grid {

  override def steps: Int = duration.getStandardHours.toInt

}

object GridBuilder {

  def grid(start:Instant, end:Instant) = GridStepEachHour(start, end)

  def gridForWork(work : Work):Grid = {
    val start = work.earliestStartTime.getOrElse(DEFAULTS.DEFAULT_EARLIEST_START)
    val end = work.latestFinishTime.getOrElse(DEFAULTS.DEFAULT_LATEST_END)
    GridStepEachHour(start, end)
  }

}